addons/mail/models/ir_websocket.py

```diff
@@ -31,35 +36,67 @@ def _update_mail_presence(self, inactivity_period):
             self.env.user if partner else guest, inactivity_period
         )
 
-    def _filter_accessible_presences(self, partners, guests):
-        """Filter presences that are accessible to current user."""
-        if self.env.user and self.env.user._is_internal():
-            return partners, guests
-        return self.env["res.partner"], self.env["mail.guest"]
-
     def _prepare_subscribe_data(self, channels, last):
         data = super()._prepare_subscribe_data(channels, last)
-        str_presence_channels = {
-            c for c in channels if isinstance(c, str) and c.startswith("odoo-presence-")
-        }
-        simplified_presence_channels = [
-            tuple(c.replace("odoo-presence-", "").split("_")) for c in str_presence_channels
-        ]
-        for channel in str_presence_channels:
+        partner_presence_data = {"ids": [], "token_by_id": {}}
+        guest_presence_data = {"ids": [], "token_by_id": {}}
+
+        for channel in channels:
+            if not isinstance(channel, str) or not channel.startswith("odoo_presence-"):
+                continue
             data["channels"].discard(channel)
-        partner_ids = [int(p[1]) for p in simplified_presence_channels if p[0] == "res.partner"]
-        partners = (
+            if not (
+                match := re.match(
+                    r"^odoo_presence-(res\.partner|mail\.guest)_(\d+)(?:-(.+))?$", channel
+                )
+            ):
+                _logger.warning("Malformed presence channel: %s", channel)
+                continue
+            model, record_id, token = match.groups()
+            record_id = int(record_id)
+            if model == "res.partner":
+                partner_presence_data["ids"].append(record_id)
+                partner_presence_data["token_by_id"][record_id] = token or ""
+            elif model == "mail.guest":
+                guest_presence_data["ids"].append(record_id)
+                guest_presence_data["token_by_id"][record_id] = token or ""
+
+        allowed_partners = (
             self.env["res.partner"]
             .with_context(active_test=False)
             .sudo()
-            .search([("id", "in", partner_ids)])
+            .search([("id", "in", partner_presence_data["ids"])])
             .sudo(False)
         )
-        guest_ids = [int(p[1]) for p in simplified_presence_channels if p[0] == "mail.guest"]
-        guests = self.env["mail.guest"].sudo().search([("id", "in", guest_ids)]).sudo(False)
-        allowed_partners, allowed_guests = self._filter_accessible_presences(partners, guests)
-        data["channels"].update((partner, "presence") for partner in allowed_partners)
-        data["channels"].update((guest, "presence") for guest in allowed_guests)
+        allowed_guests = (
+            self.env["mail.guest"]
+            .sudo()
+            .search([("id", "in", guest_presence_data["ids"])])
+            .sudo(False)
+        )
+        if not self.env.user._is_internal():
```

Comment:
Token could be used for all users/guests, we don't need to add extra conditions.

Access check should be: if valid token -> use it, if acl ok -> use it, otherwise prevent